src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_mosquitto_utils.c

```diff
@@ -0,0 +1,261 @@
+ * a specific user property with the given key and value.
+ *
+ * @param[in] props Pointer to the head of the MQTT v5 property list.
+ * @param[in] key The key of the user property to search for.
+ * @param[in] value The expected value for the given key.
+ *
+ * @return True if the property list contains the user property with the specified key and value, otherwise false.
+ *
+ * @note If there are multiple properties with the same key, this function will
+ * return true upon finding the first key-value match.
+ */
+bool ADU_mosquitto_has_user_property(const mosquitto_property* props, const char* key, const char* value)
+{
+    bool found = false;
+    while (props != NULL && !found)
+    {
+        char* k;
+        char* v;
+        const mosquitto_property* p =
+            mosquitto_property_read_string_pair(props, MQTT_PROP_USER_PROPERTY, &k, &v, false);
+        if (p != NULL && k != NULL && v != NULL)
+        {
+            found = strcmp(k, key) == 0 && strcmp(v, value) == 0;
+            free(k);
+            free(v);
+            if (!found)
+            {
+                props = mosquitto_property_next(props);
+            }
+        }
+    }
+    return found;
+}
+
+/**
+ * @brief Retrieve the value of a specific user property from an MQTT v5 property list.
+ *
+ * This function searches the provided MQTT v5 property list for a user property with the
+ * specified key. If found, it populates the value parameter with the corresponding value.
+ *
+ * @param[in] props Pointer to the head of the MQTT v5 property list.
+ * @param[in] key The key of the user property to search for.
+ * @param[out] value Pointer to the location where the function will store the corresponding
+ *                   value for the given key, if found. This value must be free'd by the caller.
+ *
+ * @return True if the property list contains the user property with the specified key,
+ *         and the value parameter has been populated. Otherwise, it returns false.
+ *
+ * @note If there are multiple properties with the same key, this function will populate
+ * the value parameter with the value of the first key it encounters.
+ *
+ * @warning If any of the input pointers are NULL, the function will log an error and return false.
+ */
+bool ADU_mosquitto_read_user_property_string(const mosquitto_property* props, const char* key, char** value)
+{
+    bool found = false;
+    if (props == NULL || value == NULL || key == NULL)
+    {
+        Log_Error("Null pointer (props=%p, value=%p, key=%p)", props, value, key);
+        return found;
+    }
+    *value = NULL;
+    while (props != NULL && !found)
+    {
+        char* k = NULL;
+        char* v = NULL;
+        const mosquitto_property* p =
+            mosquitto_property_read_string_pair(props, MQTT_PROP_USER_PROPERTY, &k, &v, false);
+        if (p != NULL && k != NULL && v != NULL)
+        {
+            found = (strcmp(k, key) == 0);
+            if (found)
+            {
+                *value = v;
```

Comment:
set v to NULL after to complete commit of move ownership. Then always free v and k at end of each iteration